crates/astria-sequencer/src/accounts/transaction.rs

```diff
@@ -0,0 +1,72 @@
+use anyhow::{
+    ensure,
+    Result,
+};
+use async_trait::async_trait;
+use serde::{
+    Deserialize,
+    Serialize,
+};
+
+use crate::{
+    accounts::{
+        state_ext::{
+            StateReadExt,
+            StateWriteExt,
+        },
+        types::{
+            Address,
+            Balance,
+            Nonce,
+        },
+    },
+    transaction::ActionHandler,
+};
+
+#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
+pub struct Transaction {
+    to: Address,
+    from: Address,
+    amount: Balance,
+    nonce: Nonce,
+}
+
+impl Transaction {
+    pub fn new(to: Address, from: Address, amount: Balance, nonce: Nonce) -> Self {
+        Self {
+            to,
+            from,
+            amount,
+            nonce,
+        }
+    }
+}
+
+#[async_trait]
+impl ActionHandler for Transaction {
+    fn check_stateless(&self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn check_stateful<S: StateReadExt + 'static>(&self, state: &S) -> Result<()> {
+        let curr_nonce = state.get_account_nonce(&self.from).await?;
+
+        // TODO: do nonces start at 0 or 1? this assumes an account's first tx has nonce 1.
+        ensure!(curr_nonce < self.nonce, "invalid nonce",);
+
+        let curr_balance = state.get_account_balance(&self.from).await?;
+        ensure!(curr_balance >= self.amount, "insufficient funds",);
+
+        Ok(())
+    }
+
+    async fn execute<S: StateWriteExt>(&self, state: &mut S) -> Result<()> {
```

Comment:
Is there anything preventing the call of `execute` without having first validated the tx? What happens if there are insufficient funds when this is called?

In general I wonder if structurally we should be moving more of the logic of "tx creates these DB level changes" directly into the state implementation.

Reply:
yeah, this function is only called inside `DeliverTx`, which always performs validation (stateless and non-stateless) before executing it. what do you mean by moving it directly into the state implementation?

Reply:
I just want to make sure we are creating hardened code here. Should it be possible to call execute before validation? If no we should enforce that it isn't possible not hope someone doesn't forget that.

Reply:
We should have a write-up about how state gets transformed when moving through these components.